src/engine/OptionalJoin.cpp

```diff
@@ -0,0 +1,372 @@
+// Copyright 2016, University of Freiburg,
+// Chair of Algorithms and Data Structures.
+// Author: Björn Buchhold (buchhold@informatik.uni-freiburg.de)
+
+#include "./OptionalJoin.h"
+
+using std::string;
+
+// _____________________________________________________________________________
+OptionalJoin::OptionalJoin(QueryExecutionContext* qec,
+                           std::shared_ptr<QueryExecutionTree> t1,
+                           bool t1Optional,
+                           std::shared_ptr<QueryExecutionTree> t2,
+                           bool t2Optional,
+                           const vector<array<size_t, 2>>& jcs) :
+  Operation(qec),
+  _joinColumns(jcs),
+  _multiplicitiesComputed(false) {
+  // Make sure subtrees are ordered so that identical queries can be identified.
+  AD_CHECK_GT(jcs.size(), 0);
+  if (t1->asString() < t2->asString()) {
+    _left = t1;
+    _leftOptional = t1Optional;
+    _right = t2;
+    _rightOptional = t2Optional;
+  } else {
+    _left = t2;
+    _leftOptional = t2Optional;
+    _right = t1;
+    _rightOptional = t1Optional;
+    for (unsigned int i = 0; i < _joinColumns.size(); i++) {
+      std::swap(_joinColumns[i][0], _joinColumns[i][1]);
+    }
+  }
+}
+
+// _____________________________________________________________________________
+string OptionalJoin::asString(size_t indent) const {
+  std::ostringstream os;
+  for (size_t i = 0; i < indent; ++i) { os << " "; }
+  os << "OPTIONAL_JOIN\n"
+     << _left->asString(indent) << " ";
+  os << "join-columns: [";
+  for (size_t i = 0; i < _joinColumns.size(); i++) {
+    os << _joinColumns[i][0] << (i < _joinColumns.size() - 1 ? " & " : "");
+  };
+  os << "]\n";
+  for (size_t i = 0; i < indent; ++i) { os << " "; }
+  os << "|X|\n"
+     << _right->asString(indent) << " ";
+  os << "join-columns: [";
+  for (size_t i = 0; i < _joinColumns.size(); i++) {
+    os << _joinColumns[i][1] << (i < _joinColumns.size() - 1 ? " & " : "");
+  };
+  os << "]";
+  return os.str();
+}
+
+// Used to generate all up to 125 combinations of left, right and result size.
```

Comment:
If we allowed C++14/17 `constexpr` with loops this could be done much nicer, right? If yes, then we probably should.

Reply:
I think you are right. I'll have to look into that later today or tomorrow.

Reply:
I did a bit of googling for the uses of constexpr in c++ 14 and 17, as I've never used it extensively before.  
As far as I could find out, there is no way, apart from the same recursion that is implemented with the templates right now, to create a constexpr loop. As the problem with finding the right template parameters is, that we need to transform a runtime value with a limited number of values into a compile time constant, I didn't find a nicer solution using constexpr, as any assignment of a constexpr would depend on a runtime value.  
Was there a specific trick you were thinking of, or is there a feature of constexpr I didn't find, that would help with these problems?